docs/melange-book/counter.md

```diff
@@ -0,0 +1,176 @@
+better way:
+
+```reason
+{counter |> Int.to_string |> React.string}
+```
+
+This uses the [pipe last operator](../communicate-with-javascript#pipe-last),
+which is useful for chaining function calls.
+
+Let's add a bit of styling to the root element of `Counter`:
+
+```reason
+<div
+  style={ReactDOMStyle.make(
+    ~padding="1em",
+    ~display="grid",
+    ~gridGap="1em",
+    ~gridTemplateColumns="25px fit-content(20px) 25px",
+    (),
+  )}>
+  <button onClick={_evt => setCounter(v => v - 1)}> {React.string("-")} </button>
+  <span> {counter |> Int.to_string |> React.string} </span>
+  <button onClick={_evt => setCounter(v => v + 1)}> {React.string("+")} </button>
+</div>
+```
+
+Unlike in React, the `style` prop in ReasonReact doesn't take a generic object,
+instead it takes an object of type `ReactDOMStyle.t` that is created by calling
+`ReactDOMStyle.make`. Obviously, this isn't an ideal way to style our app--we'll
+learn a better way to do it in a later section.
+
+Congratulations! You've created your first ReasonReact app and component. We'll
+create more complex and interesting components in future chapters.
+
+## Exercises
+
+1. There aren't any runtime errors in our app right now. But what happens if you
+   try to remove the `| None` branch of the `switch (node)` expression?
+1. What happens if you rename the `_evt` variable inside the button callback to
+   `evt`?
+1. Comment out the `[@react.component]` attribute in `Counter.re`. What happens?
+
+## Overview
+
+What we covered in this section:
+
+- How to create and run a basic ReasonReact app
+- ReasonReact components are also modules
+- OCaml has an `option` type whose value can be either `None` or `Some(_)`
+- The pipe last operator is an alternate way to invoke functions that enables
+  easy chaining of function calls
+- The `style` prop doesn't take generic objects
+
+## Solutions
+
+1. Removing the `| None` branch will result in a compilation error:
+   ```
+   Error (warning 8 [partial-match]): this pattern-matching is not exhaustive.
+   Here is an example of a case that is not matched:
+   None
+   ```
+   Basically, the compiler is telling you to handle the `None` case if you want
+   to ship your app. This is part of what makes OCaml such a type-safe language.
+1. Renaming `_evt` to `evt` results in a compilation error:
+   ```
+   Error (warning 27 [unused-var-strict]): unused variable evt.
+   ```
+   OCaml wants you to use all the variables you declare, unless they being with
+   `_` (underscore).
+1. Commenting out `[@react.component]` in `Index.re`, at the place where
+   `Counter` is used:
+   ```
+   File "src/Index.re", line 3, characters 19-27:
+   3 |   let make = () => <Counter />;
+                          ^^^^^^^^
+   Error: Unbound value Counter.makeProps
+   ```
+   For now, just remember that you need to put the `[@react.component]`
+   attribute above your `make` function if you want your component to be usable
+   in JSX. We'll explain why later.
```

Comment:
Added a reference to the PPX chapter but didn't use a link since I'm not really sure it'll be called that when I finally write it. https://github.com/melange-re/melange-re.github.io/pull/93/commits/2fbcdbeac74c78b6b871e42bab09cb7ebec86218